### PR TITLE
MAKE-1160: Use poetry, black, mypy, pydocstyle

### PR DESCRIPTION
**Pull request by dbradf**
There are no functional changes here. But there are a lot of changes. Mostly, adding docstrings, black formatting and adding type hints.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+[tool.poetry]
+name = "evergreen.py"
+version = "1.1.0"
+description = "Python client for the Evergreen API"
+authors = [
```
Comment (dbradf) [inline pyproject.toml]: To start off, I included everyone who has made a commit, in alphabetical order.

File: src/evergreen/base.py [hunk 2]
```diff
@@ -1,12 +1,20 @@
...
+from typing import Any, Callable, Dict, Optional, TYPE_CHECKING
 
+from evergreen.util import (
+    parse_evergreen_date,
+    parse_evergreen_datetime,
+    parse_evergreen_short_datetime,
+)
 
-def evg_attrib(attrib_name, type_fn=None):
+if TYPE_CHECKING:
```
Comment (dbradf) [inline src/evergreen/base.py]: This lets us import objects for type checking without creating circle dependencies.

File: mypy.ini [hunk 3]
```diff
@@ -0,0 +1,12 @@
+[mypy]
+ignore_missing_imports = True
```
Comment (ancostas) [inline mypy.ini]: Which imports do we use that can't be resolved?

File: mypy.ini [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+disallow_incomplete_defs = True
+disallow_untyped_defs = True
+disallow_untyped_calls = True
+show_error_codes = True
+
+[mypy-tests.*]
+ignore_errors = True
+
+[mypy-evergreen.cli.*]
+ignore_errors = True
```
Comment (ancostas) [inline mypy.ini]: Why? Since no one will consume the APIs here programmatically?

File: src/evergreen/base.py [hunk 5]
```diff
@@ -58,29 +66,33 @@ def evg_date_attrib(attrib_name):
 class _BaseEvergreenObject(object):
     """Common evergreen object."""
 
-    def __init__(self, json, api):
-        """
-        Create an instance of an evergreen task.
-        """
+    def __init__(self, json: Dict, api: EvergreenApi) -> None:
```
Comment (ancostas) [inline src/evergreen/base.py]: IMO we should do a find and replace: `json: Dict` --> `json: Dict[str, Any]`

File: .evergreen.yml [hunk 6]
```diff
@@ -57,13 +29,12 @@ functions:
...
           . venv/bin/activate
 
-
           if [ "${is_patch}" = "true" ]; then
             # Do not deploy on patches.
             exit 0
           fi
 
-          scripts/deploy.sh ${pypi_user} ${pypi_password}
+          poetry publish --username ${pypi_user} --password ${pypi_password}
```
Comment (ancostas) [inline .evergreen.yml]: Does this prevent overwriting existing versions?

Comment (dbradf) [reply .evergreen.yml]: It does not. I haven't been able to find a good way to do that with any of the python publishing tools without manually writing something around them.

File: mypy.ini (hunk 3 above)
Comment (dbradf) [reply mypy.ini]: 3rd party libraries:
```
12: error: Cannot find implementation or library stub for module named 'structlog'  [import]
12: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
13: error: Cannot find implementation or library stub for module named 'structlog.stdlib'  [import]
14: error: Cannot find implementation or library stub for module named 'tenacity'  [import]
40: error: Cannot find implementation or library stub for module named 'urlparse'  [import]
```

File: mypy.ini (hunk 4 above)
Comment (dbradf) [reply mypy.ini]: We haven't really supported the cli code for this. It is mostly just used to manually test things and isn't really the purpose of this repository. As such, it is a lot of work to bring it up to standards. So, for now I think we should just ignore it and if someone wants to go clean it up at some point they can.

File: src/evergreen/base.py (hunk 5 above)
Comment (dbradf) [reply src/evergreen/base.py]: done.

File: .evergreen.yml [hunk 7]
```diff
@@ -57,13 +29,12 @@ functions:
           set -o errexit
           . venv/bin/activate
 
-
           if [ "${is_patch}" = "true" ]; then
             # Do not deploy on patches.
             exit 0
           fi
 
-          scripts/deploy.sh ${pypi_user} ${pypi_password}
```
Comment (ancostas) [inline .evergreen.yml]: I think we can get rid of `deploy.sh` now.

**Comment by dbradf** [comment]
evergreen merge

**Pull request merged**